tt_metal/tt_stl/any_range.hpp

```diff
@@ -651,95 +699,135 @@ class AnyRangeAdaptor<TAnyRange, TRange, TRangeTag, std::enable_if_t<is_tag_v<TR
 
    private:
     template <class TRange, class TEnable = void>
-    struct enable_if_range_compatible : std::enable_if<detail::range_compatible<TRange, reference, range_category>> {};
+    struct enable_if_range_compatible : std::enable_if<
+                                            detail::range_compatible<TRange, reference, range_category>,
+                                            detail::FinalAnyRangeAdaptor<BasicAnyRange, TRange>> {};
 
     // Don't hide copy and move constructors
     template <class TRange>
     struct enable_if_range_compatible<
         TRange,
         std::enable_if_t<std::is_same_v<detail::remove_cvref_t<TRange>, BasicAnyRange>>> {};
 
+    template <class TAdaptor>
+    struct enable_if_storage_compatible
+        : std::enable_if<
+              sizeof(BasicAnyRange) >= sizeof(TAdaptor) and detail::is_aligned_to_v<TAdaptor, alignof(BasicAnyRange)>,
+              TAdaptor> {};
+
+    template <class TIterator>
+    struct enable_if_iterator_compatible : std::enable_if<std::is_constructible_v<iterator, TIterator>> {};
+
    public:
-    template <class TRange, class = typename enable_if_range_compatible<TRange>::type>
+    /**
+     * @brief Constructor from forwarding reference of underlying range type
+     * @param range The cvref-qualified range type to erase; lvalues are stored as reference, rvalues are
+     * move-constructed into storage, transferring ownership.
+     */
+    template <
+        class TRange,
+        class TAdaptor = typename enable_if_range_compatible<TRange>::type,
+        class = typename enable_if_storage_compatible<TAdaptor>::type,
+        class = typename enable_if_iterator_compatible<detail::iterator_t<TRange>>::type>
     // NOLINTNEXTLINE(bugprone-forwarding-reference-overload)
     BasicAnyRange(TRange &&range) {
-        using final_adaptor_type = detail::FinalAnyRangeAdaptor<BasicAnyRange, TRange>;
+        ::new (static_cast<void *>(bytes)) TAdaptor(static_cast<TRange &&>(range));
+    }
 
-        static_assert(sizeof(BasicAnyRange) >= sizeof(final_adaptor_type));
-        static_assert(alignof(BasicAnyRange) % alignof(final_adaptor_type) == 0);
+#ifdef __has_cpp_attribute
+#if __has_cpp_attribute(clang::lifetimebound)
 
-        ::new (static_cast<void *>(bytes)) final_adaptor_type(static_cast<TRange &&>(range));
+    /**
+     * @brief Higher-priority constructor from lvalue reference of underlying range type
+     * @note Diagnoses attempts to type-erase a dangling reference if lifetimebound attribute is supported.
+     * @param range The lvalue reference range type to erase; stored as reference.
+     */
+    template <
+        class TRange,
+        class TAdaptor = typename enable_if_range_compatible<TRange &>::type,
+        class = typename enable_if_storage_compatible<TAdaptor>::type,
+        class = typename enable_if_iterator_compatible<detail::iterator_t<TRange>>::type>
+    BasicAnyRange(TRange &range [[clang::lifetimebound]]) {
```

Comment:
It seems like this constructor would be disabled if not `__has_cpp_attribute(clang::lifetimebound)`. Was it the intention? Wouldn't we want to have the same constructor available without the lifetimebound annotation if the feature is not supported?

Reply:
It is the intention, yes. The forwarding constructor can pick up lvalues if this overload is not defined, and when lifetimebound attribute is not available, compilation errors for invalid constructor calls will have one less overload to list which should make reading the errors a little easier.